File: JavaSource/src/main/java/ru/akhcheck/patterns/memento/Originator.java
package ru.akhcheck.patterns.memento;

import java.util.ArrayList;
import java.util.List;

/**
 * Хозяин состояния: игра.
 *
 * <p>Только он умеет делать снимок и только он умеет из снимка восстановиться.
 * Смотритель снаружи не может ни прочитать поля, ни собрать снимок сам.
 */
public class Originator {
    private static final int LEVEL_EXPERIENCE = 100;

    private State state = State.empty();

    public void startNewGame() {
        state = new State(1, 0, List.of("knife"));
    }

    public void increaseLevel() {
        int nextLevel = state.getLevel() + 1;
        List<String> inventory = new ArrayList<>(state.getInventory());
        inventory.add("trophy-" + nextLevel);
        state = new State(nextLevel, state.getExperience() + LEVEL_EXPERIENCE, inventory);
    }

    public Memento save(String saveName) {
        return new Memento(saveName, state);
    }

    public void restore(Memento memento) {
        state = memento.getState();
    }

    public void showCurrentProgress() {
        System.out.println(
                "level=" + state.getLevel()
                        + " experience=" + state.getExperience()
                        + " inventory=" + state.getInventory()
        );
    }
}

File: Memento/python-source/Originator.py
from __future__ import annotations

from Memento import Memento
from Memento import State


class Originator:
    """Хозяин состояния: игра.

    Только он умеет делать снимок и только он умеет из снимка восстановиться.
    Смотритель снаружи не может ни прочитать поля, ни собрать снимок сам.
    """

    LEVEL_EXPERIENCE = 100

    def __init__(self) -> None:
        self._state = State()

    def start_new_game(self) -> None:
        self._state = State(level=1, experience=0, inventory=("knife",))

    def increase_level(self) -> None:
        self._state = State(
            level=self._state.level + 1,
            experience=self._state.experience + self.LEVEL_EXPERIENCE,
            inventory=(*self._state.inventory, f"trophy-{self._state.level + 1}"),
        )

    def save(self, save_name: str) -> Memento:
        return Memento(name=save_name, state=self._state)

    def restore(self, memento: Memento) -> None:
        self._state = memento.state

    def show_current_progress(self) -> None:
        print(
            f"level={self._state.level} "
            f"experience={self._state.experience} "
            f"inventory={list(self._state.inventory)}"
        )
